### PR TITLE
🚨 [DiffusionGemma] Return router logits and load balancing loss

### PR DESCRIPTION
**Pull request by kashif**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46642)

DiffusionGemma is a MoE model but it never surfaced the router logits, so downstream trainers couldn't add a load balancing loss on the experts.

This wires it up the same way as the other MoE models (qwen3_moe as reference):
- adds `output_router_logits` and `router_aux_loss_coef` to the text config
- the decoder router now returns its raw logits, which the recorder exposes as `router_logits`
- `DiffusionGemmaForBlockDiffusion` returns `router_logits` and the load balancing `aux_loss` when the flag is set

It also makes `gemma4` and `gemma4_unified` trainable with the MoE load balancing loss (they were missing it too), and diffusion gemma inherits the config and imports the loss func from mixtral through modular.

The loss is only over the decoder experts (the canvas denoising path, which is unpadded so no mask is needed). Off by default, so generation is unchanged.

Note: the gemma routers now return their raw logits as the first output instead of the post-softmax probabilities. The model output is identical (the decoder layer discards that value and the router module isn't exported), but if you import a router `Module` directly, its first return value changed.

Tests: added `test_output_router_logits` and a general `test_load_balancing_loss` in `test_modeling_common.py`. Full diffusion_gemma, gemma4 and gemma4_unified suites pass on CPU.

AI assistance was used.

### CODE REVIEW

File: src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py [hunk 1]
```diff
@@ -85,10 +85,17 @@ class DiffusionGemmaTextConfig(Gemma4TextConfig):
...
+        Whether or not to return the router logits of the MoE layers, used to compute the load balancing loss when
+        training the experts.
+    router_aux_loss_coef (`float`, *optional*, defaults to 0.001):
+        The coefficient for the load balancing loss added to the main loss.
     """
 
     model_type = "diffusion_gemma_text"
     final_logit_softcapping = 30.0
+    output_router_logits: bool = False
+    router_aux_loss_coef: float = 0.001
```
Comment (zucchini-nlp) [inline src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py]: huh, was also missing in gemma4 then. Do you mind making gemma4 trainable with MoE loss as well, so it is then copied here by modular?

File: src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py [hunk 2]
```diff
@@ -1247,16 +1255,102 @@ def create_diffusion_decoder_attention_mask(
         return {"full_attention": full_mask, "sliding_attention": sliding_mask}
 
 
+def load_balancing_loss_func(
```
Comment (zucchini-nlp) [inline src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py]: looks same as in all MoE models. We can import from another model, modular should know how to copy it

**Review by zucchini-nlp — COMMENTED** [review]
Nice, lets sync with gemma4 also

File: src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py (hunk 2 above)
Comment (kashif) [reply src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py]: ok sure!

File: src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py (hunk 1 above)
Comment (kashif) [reply src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py]: Done. Added the MoE load balancing loss to gemma4 itself (and gemma4 unified), so diffusion gemma now inherits the config and wiring through modular.

File: src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py (hunk 2 above)
Comment (kashif) [reply src/transformers/models/diffusion_gemma/modular_diffusion_gemma.py]: Switched to importing it from mixtral, modular copies it into the generated file. Same now in gemma4.

**Comment by kashif** [comment]
Reworked as discussed: gemma4 (and gemma4 unified) are now trainable with the MoE load balancing loss, and the load balancing func is imported from mixtral instead of duplicated. Diffusion gemma inherits the config and the func through modular, so its generated files are unchanged. gemma4 / gemma4 unified / diffusion gemma tests pass (638), repo consistency is green.

File: src/transformers/models/gemma4/modular_gemma4.py
Comment (zucchini-nlp) [inline src/transformers/models/gemma4/modular_gemma4.py]: we can inherit from `MoeModelOutputWithPast` which already has router-logits :)

File: src/transformers/models/gemma4/modular_gemma4.py [hunk 3]
```diff
@@ -1128,7 +1139,8 @@ def forward(self, hidden_states: torch.Tensor) -> tuple[torch.Tensor, torch.Tens
         # Apply per-expert scale directly to the weights
         top_k_weights = top_k_weights * self.per_expert_scale[top_k_index]
 
-        return router_probabilities, top_k_weights, top_k_index
+        # The first element is recorded as `router_logits` for the load balancing loss
+        return expert_scores, top_k_weights, top_k_index
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma4/modular_gemma4.py:1208]: kinda breaking tho I don't mind it to get a trainable MoE. Let's mark PR with a single 🚨

File: src/transformers/models/gemma4/modular_gemma4.py
Comment (zucchini-nlp) [inline src/transformers/models/gemma4/modular_gemma4.py]: same here, could be `MoeCausalLMOutputWithPast`

File: tests/models/diffusion_gemma/test_modeling_diffusion_gemma.py [hunk 4]
```diff
@@ -249,6 +249,25 @@ def test_disk_offload_safetensors(self):
         pass
 
     # Tests designed specifically for `DiffusionGemma`, excluding generation tests.
+    def test_output_router_logits(self):
```
Comment (zucchini-nlp) [inline tests/models/diffusion_gemma/test_modeling_diffusion_gemma.py:300]: a general test would be nice in `test_modeling_common.py`

Such that we model is MoE and is generative (in AutoCausalLM or AutoMultimodalLM), we expect the output to have a non-zero field `aux_loss`. That would enforce a better standard with MoE

**Review by zucchini-nlp — COMMENTED** [review]
Thanks a lot, can we add one test to see if there are any other models with missing aux-loss and we'll merge

File: src/transformers/models/gemma4/modular_gemma4.py (hunk 3 above)
Comment (kashif) [reply src/transformers/models/gemma4/modular_gemma4.py:1208]: I don't think it actually breaks anything. The first router output is dropped by the layer (`_, top_k_weights, top_k_index = self.router(...)`) and the module isn't exported, so the model output is identical. router_logits is just a new field. Skip the 🚨?

File: src/transformers/models/gemma4/modular_gemma4.py
Comment (kashif) [reply src/transformers/models/gemma4/modular_gemma4.py]: Kept it additive on purpose. Inheriting MoeModelOutputWithPast moves router_logits ahead of shared_kv_states and changes the to_tuple order, appending keeps the existing indices stable. Can switch if you prefer though.

Comment (kashif) [reply src/transformers/models/gemma4/modular_gemma4.py]: Same here, inheriting it would push aux_loss before logits and shift the existing fields in to_tuple, so I appended instead to keep them put.

File: tests/models/diffusion_gemma/test_modeling_diffusion_gemma.py (hunk 4 above)
Comment (kashif) [reply tests/models/diffusion_gemma/test_modeling_diffusion_gemma.py:300]: Added test_load_balancing_loss in common. It asserts a non-zero aux_loss for generative MoE models that expose one, and skips the ones that carry output_router_logits but don't compute it so they don't go red. Checked mixtral, qwen3_moe, gemma4 and diffusion_gemma pass, llama skips.

File: src/transformers/models/gemma4/modular_gemma4.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/gemma4/modular_gemma4.py:1208]: yeah, I see your point, globally non very breaking. My concern was that if users import a single Module, the first output won't be the same anymore

In any case, we have it in PR description as well

**Review by zucchini-nlp — APPROVED** [review]
Nice, thanks a lot! I think merging is broken for now due to CI updates, we can merge when it's fixed

File: src/transformers/models/gemma4/modular_gemma4.py (hunk 3 above)
Comment (kashif) [reply src/transformers/models/gemma4/modular_gemma4.py:1208]: Fair point on the single-module import. Added a note about it to the PR description.

**Comment by zucchini-nlp** [comment]
The failing test in CI looks relevant here, it is the newly added testcase 👀

**Comment by zucchini-nlp** [comment]
@bot /style
